src/Chainweb/BlockHeader/Validation.hs

```diff
@@ -414,67 +690,61 @@ prop_block_featureFlags b
 
-prop_block_chainwebVersion
-    :: ParentHeader
-        -- ^ parent header of the checked block header
-    -> BlockHeader
-        -- ^ the block header that is checked
-    -> Bool
-prop_block_chainwebVersion (ParentHeader p) b =
+prop_block_chainwebVersion :: ChainStep -> Bool
+prop_block_chainwebVersion (ChainStep (ParentHeader p) b) =
     _blockChainwebVersion p == _blockChainwebVersion b
 
-prop_block_creationTime
-    :: ParentHeader
-        -- ^ parent header of the checked block header
-    -> BlockHeader
-        -- ^ the block header that is checked
-    -> Bool
-prop_block_creationTime (ParentHeader p) b
+prop_block_weight :: ChainStep -> Bool
+prop_block_weight (ChainStep (ParentHeader p) b)
+    | isGenesisBlockHeader b = _blockWeight b == _blockWeight p
+    | otherwise = _blockWeight b == expectedWeight
+  where
+    expectedWeight = int (targetToDifficulty (_blockTarget b)) + _blockWeight p
+
+prop_block_chainId :: ChainStep -> Bool
+prop_block_chainId (ChainStep (ParentHeader p) b)
+    = _blockChainId p == _blockChainId b
+
+-- -------------------------------------------------------------------------- --
+-- Mutli chain inductive properties
+
+prop_block_epoch :: WebStep -> Bool
+prop_block_epoch (WebStep _as (ChainStep p b))
+    = _blockEpochStart b == epochStart p (_blockCreationTime b)
+
+prop_block_creationTime :: WebStep -> Bool
+prop_block_creationTime (WebStep _as (ChainStep (ParentHeader p) b))
     | isGenesisBlockHeader b = _blockCreationTime b == _blockCreationTime p
     | otherwise = _blockCreationTime b > _blockCreationTime p
 
-prop_block_weight
-    :: ParentHeader
-        -- ^ parent header of the checked block header
-    -> BlockHeader
-        -- ^ the block header that is checked
-    -> Bool
-prop_block_weight (ParentHeader p) b
-    | isGenesisBlockHeader b = _blockWeight b == _blockWeight p
-    | otherwise = _blockWeight b == int (targetToDifficulty (_blockTarget b)) + _blockWeight p
-
-prop_block_chainId
-    :: ParentHeader
-        -- ^ parent header of the checked block header
-    -> BlockHeader
-        -- ^ the block header that is checked
-    -> Bool
-prop_block_chainId (ParentHeader p) b
+prop_block_adjacent_chainId :: WebStep -> Bool
+prop_block_adjacent_chainId (WebStep _as (ChainStep (ParentHeader p) b))
     = _blockChainId p == _blockChainId b
+
+-- | TODO: we don't current check this here. It is enforced in the cut merge
+-- algorithm , namely in 'monotonicCutExtension'. The property that is checked
+-- in the cut validation is stronger than the braiding property that we could
+-- check here (which is the property that is described in the chainweb paper).
+-- So, I check here would add some redundancy, but it's not clear of how much
+-- actual value that would be.
+--
+-- If we want to enforce it here, too, we would have to look back 2 steps in
+-- history.
+--
+prop_block_braiding :: WebStep -> Bool
+prop_block_braiding _ = True
```

Comment:
Remove if not checked?

Reply:
I kept it because of the comment above. It would be easy to include all grandparents into a WebStep.

Performance with it shouldn't make a big difference here in validation. The size of a WebStep is `O(diameter graph) == O(log(number_of_chains))` in any case. But it might affect performance of database lookups (because caches would grow by `O(order graph) == O(number_of_chains)` with respect to the depth in history. Although I am not sure how much it matters. It would also add overhead in creating test cases. 

Given that it's an redundant check, I am not sure if the added robustness pays off.